### PR TITLE
feat(core) explict flag for regex route.path

### PR DESCRIPTION
**Pull request by Suika-Kong**
Another solution for #9022.

fix CT-344

We need to remember to add migration, and also "back migration" for EE.

**PLEASE MERGE AFTER: #9024**

### CODE REVIEW

File: kong/db/schema/typedefs.lua [hunk 1]
```diff
@@ -449,15 +452,15 @@ local function validate_path_with_regexes(path)
   -- We can't take an ok from validate_path as a success just yet,
   -- because the router is currently more strict than RFC 3986 for
   -- non-regex paths:
-  if ngx.re.find(path, [[^[a-zA-Z0-9\.\-_~/%]*$]]) then
+  if path:sub(1,2) ~= "~*" then
```
Comment (chronolaw) [inline kong/db/schema/typedefs.lua:455]: should have a space, like `sub(1, 2)`

File: kong/router.lua [hunk 2]
```diff
@@ -425,8 +425,9 @@ local function marshall_route(r)
       match_weight = match_weight + 1
       for i = 1, count do
         local path = paths[i]
+        local is_regex = path:sub(1,2) == "~*"
```
Comment (chronolaw) [inline kong/router.lua:428]: should have a space, like `sub(1, 2)`

File: spec/02-integration/05-proxy/02-router_spec.lua [hunk 3]
```diff
@@ -2191,6 +2191,12 @@ for _, strategy in helpers.each_strategy() do
                 }))
 
                 local data = assert.response(res).has.jsonbody()
+                if not data.vars then
```
Comment (chronolaw) [inline spec/02-integration/05-proxy/02-router_spec.lua:2194]: Is this a debug code?

Comment (Suika-Kong) [reply spec/02-integration/05-proxy/02-router_spec.lua:2194]: You're right.

Comment (chronolaw) [reply spec/02-integration/05-proxy/02-router_spec.lua]: Should we remove them?

File: CHANGELOG.md [hunk 4]
```diff
@@ -128,6 +128,9 @@
   [rfc3986](https://datatracker.ietf.org/doc/html/rfc3986#section-2.2),
   we should write all other characters without percent-encoding.
   [#9024](https://github.com/Kong/kong/pull/9024)
+- Use `"~*"` as prefix to indicate a `router.path` is a regex pattern. We no longer guess
```
Comment (chronolaw) [inline CHANGELOG.md:131]: maybe ' the `router.path`' ?

**Review by chronolaw — APPROVED** [review]

**Review by bungle — DISMISSED** [review]
It is wrong to use `~*` prefix and do case-sensitive matching. The `~` should be case-sensitive match. See: https://github.com/Kong/kong/blob/master/kong/router.lua#L960

We only add `ajo` and not `i` modifier.

**Comment by Suika-Kong** [comment]
> NOTE: this will break all existing:
> 
> 1. `scripts` that call admin apis and want regexes.
> 2. `yaml` (and other config files) that are used to configure dbless, or import data into control plane, traditional node, or imported by deck.
> 3. `kong ingress controller` might also be affected
> 
> In that sense I think this is a too big of a change that comes too close to `3.0.0` release. That said, `major` version is the version to do these kind of changes. It can be a can of worms. @dndx could you comment on this?

Right. I will write a document about that to notify the people involved.

File: CHANGELOG.md [hunk 5]
```diff
@@ -129,6 +129,9 @@
   [rfc3986](https://datatracker.ietf.org/doc/html/rfc3986#section-2.2),
   we should write all other characters without percent-encoding.
   [#9024](https://github.com/Kong/kong/pull/9024)
+- Use `"~*"` as prefix to indicate a `route.path` is a regex pattern. We no longer guess
```
Comment (chronolaw) [inline CHANGELOG.md:132]: Now we should use `~` instead of `~*`?

**Comment by Suika-Kong** [comment]
I've tested migration in my local env and it works as expected.

**Review by fffonion — APPROVED** [review]

File: kong/db/schema/typedefs.lua [hunk 6]
```diff
@@ -449,10 +452,12 @@ local function validate_path_with_regexes(path)
   -- We can't take an ok from validate_path as a success just yet,
   -- because the router is currently more strict than RFC 3986 for
   -- non-regex paths:
```
Comment (bungle) [inline kong/db/schema/typedefs.lua:454]: Is this comment relevant anymore or perhaps wrong?

File: kong/db/schema/typedefs.lua [hunk 7]
```diff
@@ -449,10 +452,12 @@ local function validate_path_with_regexes(path)
...
-  if ngx.re.find(path, [[^[a-zA-Z0-9\.\-_~/%]*$]]) then
+  if path:sub(1, 1) ~= "~" then
     return true
   end
 
+  path = path:sub(2)
+
   -- URI contains characters outside of the list recognized by the
   -- router as valid non-regex paths.
```
Comment (bungle) [inline kong/db/schema/typedefs.lua:462]: Same here.

File: kong/router.lua [hunk 8]
```diff
@@ -425,8 +425,9 @@ local function marshall_route(r)
       match_weight = match_weight + 1
       for i = 1, count do
         local path = paths[i]
+        local is_regex = path:sub(1, 1) == "~"
```
Comment (bungle) [inline kong/router.lua:428]: `sub` is already localized so:

```suggestion
        local is_regex = sub(path, 1, 1) == "~"
```

File: kong/router.lua [hunk 9]
```diff
@@ -440,6 +441,7 @@ local function marshall_route(r)
 
         else
 
+          path = path:sub(2)
```
Comment (bungle) [inline kong/router.lua:444]:
```suggestion
          path = sub(path, 2)
```

**Review by bungle — CHANGES_REQUESTED** [review]

File: kong/db/schema/typedefs.lua (hunk 6 above)
Comment (Suika-Kong) [reply kong/db/schema/typedefs.lua:454]: You are right. It doesn't hold now.

File: kong/db/schema/typedefs.lua [hunk 10]
```diff
@@ -295,7 +295,10 @@ typedefs.port = Schema.define {
 
 typedefs.path = Schema.define {
   type = "string",
-  starts_with = "/",
+  match_any = {
+    patterns = {"^/", "^~/"},
+    err = "should start with: /",
```
Comment (kikito) [inline kong/db/schema/typedefs.lua:300]:
```suggestion
    err = "should start with: / (fixed path) or ~/ (regex path)",
```

File: kong/db/schema/typedefs.lua [hunk 11]
```diff
@@ -446,15 +449,12 @@ local function validate_path_with_regexes(path)
     return ok, err, err_code
   end
 
-  -- We can't take an ok from validate_path as a success just yet,
-  -- because the router is currently more strict than RFC 3986 for
-  -- non-regex paths:
-  if ngx.re.find(path, [[^[a-zA-Z0-9\.\-_~/%]*$]]) then
+  if path:sub(1, 1) ~= "~" then
```
Comment (kikito) [inline kong/db/schema/typedefs.lua:452]: Could this be made backwards compatible by extending this with [#9022](https://github.com/Kong/kong/pull/9022) (keeping the heuristic alive for 3.0, but deprecated)?

```suggestion
  if path:sub(1, 1) ~= "~" and is_not_regex(path) then
```

File: kong/db/schema/typedefs.lua (hunk 10 above)
Comment (kikito) [inline kong/db/schema/typedefs.lua:300]:
```suggestion
    err = "should start with: / (fixed path) or ~/ (regex path)",
```

File: kong/router.lua [hunk 12]
```diff
@@ -425,8 +425,9 @@ local function marshall_route(r)
       match_weight = match_weight + 1
       for i = 1, count do
         local path = paths[i]
+        local is_regex = sub(path, 1, 1) == "~"
```
Comment (kikito) [inline kong/router.lua:428]:
```suggestion
        local is_regex = sub(path, 1, 1) == "~" or is_regex(path)
```

**Comment by Suika-Kong** [comment]
> Could this be made backwards compatible by extending this with [#9022](https://github.com/Kong/kong/pull/9022) (keeping the heuristic alive for 3.0, but deprecated)?

@kikito I've discussed this with @dndx and we decided not to apply this solution.
We will have the traditional format of route coexist with ATC at least for 3.X, and the problem we want to address with those 2 PRs are important. And @dndx thinks this solution introduces more complexity.

**Pull request merged**